prep release: v1.34.1 (#4245)
apollographql/router: CHANGELOG.md

This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

# [1.34.1] - 2023-11-21

## 🐛 Fixes

### Authorization: Fix fragment filtering ([Issue #4060](https://github.com/apollographql/router/issues/4060))

If a fragment was removed, whether because its condition cannot be fulfilled or its selections were removed, then the corresponding fragment spreads must be removed from the filtered query.

This also fixes the error paths related to fragments: before, the path started at the fragment definition, while now the fragment's errors are added at the point of application

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/4155

### special case for __typename in authorization ([PR #3821](https://github.com/apollographql/router/pull/3821))

When evaluating authorization directives on fields returning interfaces,
we require the usage of fragments with type conditions if the interface
implementors have different security requirements. For the __typename
field though, we must make an exception, because it should be available

> for all implementors

Reviewer: formatting: merge these lines.

Replace more futures channels with tokio channels to ensure that channel bounds are respected as expected.

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4138

### Ensure that license_stream and files bounded channels are bounded ([Issue #4109](https://github.com/apollographql/router/issues/4109)), ([Issue #4110](https://github.com/apollographql/router/issues/4110))

Convert `futures` channels to `tokio` channels to ensure that channel bounds are correctly observed.

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4111

### Enfore JWT expiration for subscriptions ([Issue #3947](https://github.com/apollographql/router/issues/3947))

If a JWT expires whilst a subscription is executing, the subscription should be terminated. This also applies to deferred responses.

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4166

### Convert notification from futures channels to tokio channels ([Issue #4117](https://github.com/apollographql/router/issues/4117))

Convert `futures` channels to `tokio` channels to ensure that channel bounds are correctly observed.

> 

Reviewer: Could probably merge these all together into one entry.

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4118

### Use less recursion in GraphQL parser ([Issue #4142](https://github.com/apollographql/router/issues/4142))

This updates the Router with the latest `apollo-parser` version,
which removes unnecessary use of recursion for parsing repeated syntax elements
such as enum values and union members in type definitions.

> Some documents that used to hit the parser’s recursion limit will now successfully parse.

Reviewer: Same formatting comment as above: merge these lines?

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/4144

### Port to apollo-compiler 1.0 beta ([PR #4038](https://github.com/apollographql/router/pull/4038))

Version 1.0 is a near-complete rewrite of `apollo-compiler`.
Using it in the Router unblocks a lot of upcoming work.

As a more immediate benefit, some serialization-related bugs including 
[Issue #3541](https://github.com/apollographql/router/issues/3541) are fixed.
The representation of GraphQL documents in previous compiler versions was immutable.
When modifying a query (such as to remove `@authenticated` fields from an unauthenticated request)
the Router would build a new data structure with `apollo-encoder`, serialize it, and reparse it.
`apollo-compiler`` 1.0 allows mutating a document in-place, skipping the serialization step entirely.

By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/4038

### Propagate multi-value headers to subgraphs ([Issue #4153](https://github.com/apollographql/router/issues/4153))

> Use `HeaderMap.append` instead of `insert` to prevent erasing previous values when using multiple headers with the same name.

Reviewer: ```suggestion
Use `HeaderMap.append` instead of `insert` to avoid erasing previous values when using multiple headers with the same name.
```

## 📃 Configuration

### Adds a `poll_interval` option to the JWKS endpoint in the Authentication plugin ([Issue #4185](https://github.com/apollographql/router/issues/4185))

In order to avoid rate limiting concerns on JWKS endpoints, this introduces a new `poll_interval` configuration option to set the interval per each JWKS URL. The configuration option takes in a human-readable duration (e.g. `60s` or `1minute 30s`). Example that sets interval to 30 seconds: 

```yml
authentication:
  router:
    jwt:
      jwks:
        - url: https://dev-zzp5enui.us.auth0.com/.well-known/jwks.json
          poll_interval: 30s
```

By [@lleadbet](https://github.com/lleadbet) in https://github.com/apollographql/router/pull/4212

### Configuration: allow custom health check endpoint ([Issue #2938](https://github.com/apollographql/router/issues/2938))

Adds a configuration option for custom health check endpoints, `health_check.path`, with `/health` as the default value.

By [@aaronArinder](https://github.com/aaronArinder) in https://github.com/apollographql/router/pull/4145

### Configurable JWKS Poll Interval ([Issue #4185](https://github.com/apollographql/router/issues/4185))

> The poll interval was previously hardcoded to 60 seconds. It is still the default now, but can be configured through the new `poll_interval` configuration option under each JWKS entry to avoid becoming rate-limited per endpoint.

Reviewer: ```suggestion
The poll interval was previously hardcoded to 60 seconds. That remains the default, but can now be configured through the new `poll_interval` configuration option under each JWKS entry to avoid becoming rate-limited per endpoint.
```

By [@lleadbet](https://github.com/lleadbet) in https://github.com/apollographql/router/pull/4212

## 📚 Documentation

### docs: clarify coprocessor router stage responses ([PR #4189](https://github.com/apollographql/router/pull/4189))

The coprocessor RouterRequest and RouterResponse stages fully support `control: { break: 500 }` but the response body must be a string. This change provides examples in the "Terminating a client request" section.

<!-- start metadata -->

> ---

Reviewer: Delete
